File: python/sieve.py
def primes(limit):
    if limit < 2:
        return []

    lst = [1] * (limit + 1)

    for i in range(4, limit + 1, 2):
        lst[i] = 0

    for i in range(3, limit + 1, 2):
        k = i + i
        while k <= limit:
            lst[k] = 0
            k += i

    return [i + 2 for (i, n) in enumerate(lst[2:]) if n == 1]

File: elixir/sieve.ex
defmodule Sieve do
  @doc """
  Generates a list of primes up to a given limit.
  """
  @spec primes_to(non_neg_integer) :: [non_neg_integer]
  def primes_to(limit) when limit < 2, do: []
  def primes_to(limit) do
    st = MapSet.new(2..limit)
    2..limit
    |> Enum.reduce(st, fn n, acc ->
      if MapSet.member?(acc, n) do
        (n+n)..limit//n
        |> Enum.reduce(acc, fn x, acc -> MapSet.delete(acc, x) end)
      else
        acc
      end
    end)
    |> Enum.sort()
  end
end

